### PR TITLE
[OpenTracing] Fix context propagation between OpenTracing and Datadog

### PR DESCRIPTION
**Pull request by delner**
When using `ddtrace` OpenTracer in conjunction with an app already instrumented by `ruby-rack-tracer`, we were seeing some disjoint traces. Spans generated by `ddtrace` out of the box instrumentation was not ending up parented by the OpenTracing integration parent span (in this case, Rack), resulting in multiple traces that should have appeared as one.

To fix this, the pull request does two things:

 - For distributed tracing oriented OT integrations, they pass `SpanContext` objects as `child_of`. We need to make sure the `Datadog::Context` object that comes from this is set on the tracer, such that the following generated Datadog span inherits the proper trace info (trace ID, parent ID.) It was not actually using this context properly before.
 - For the first OpenTracing span created, regardless of being preceded or followed by Datadog spans, it was not resolving a Datadog context (because it was trying to derive it from a non-existing OT parent span), and thus passing `nil` to the `start_span` method which would create a brand new context that would not be re-used on subsequent spans. We now make sure its passing a `Datadog::Context` to `start_span`.

### CODE REVIEW

File: lib/ddtrace/opentracer/tracer.rb [hunk 1]
```diff
@@ -56,6 +56,28 @@ def start_active_span(operation_name,
...
+        # the benefit of any out-of-the-box tracing from Datadog preceding
+        # the OpenTracer::Tracer.
+        #
+        # We do this here instead of in #start_span because #start_span generates
+        # spans that are not assigned to a scope, a.k.a not supposed to be used by
+        # subsequent spans implicitly. By using the existing Datadog context, the span
+        # effectively ends up "assigned to a scope", by virtue of being added to the
+        # Context. Hence, it would behave more like an active span, which is why it
+        # should only be here.
+        unless child_of || ignore_active_scope
```
Comment (delner) [inline lib/ddtrace/opentracer/tracer.rb]: This block is responsible for allowing OpenTracer spans to append to traces that were created by Datadog auto-instrumentation. e.g. `datadog.span` --> `open_tracer.span`.

File: lib/ddtrace/opentracer/tracer.rb [hunk 2]
```diff
@@ -65,6 +87,12 @@ def start_active_span(operation_name,
           ignore_active_scope: ignore_active_scope
         )
 
+        # Overwrite the tracer context with the OpenTracing managed context.
+        # This is mostly for the benefit of any out-of-the-box tracing from Datadog,
+        # such that spans generated by that tracing will be attached to the OpenTracer
+        # parent span.
+        datadog_tracer.provider.context = span.datadog_span.context
```
Comment (delner) [inline lib/ddtrace/opentracer/tracer.rb]: This line is responsible for allowing Datadog auto-instrumentation spans to append to traces that were created by OpenTracer. e.g. `open_tracer.span` --> `datadog.span` .

File: lib/ddtrace/opentracer/tracer.rb [hunk 3]
```diff
@@ -166,6 +190,19 @@ def extract(format, carrier)
           nil
         end
       end
+
+      private
+
+      def inherited_span_context(parent, ignore_active_scope: false)
```
Comment (delner) [inline lib/ddtrace/opentracer/tracer.rb]: This is just a refactor from `#start_span`.

**Comment by delner** [comment]
Made some significant changes to how context is resolved and written to. This should be ready for review.

File: lib/ddtrace/opentracer/tracer.rb [hunk 4]
```diff
@@ -56,6 +56,28 @@ def start_active_span(operation_name,
...
+        # spans that are not assigned to a scope, a.k.a not supposed to be used by
+        # subsequent spans implicitly. By using the existing Datadog context, the span
+        # effectively ends up "assigned to a scope", by virtue of being added to the
+        # Context. Hence, it would behave more like an active span, which is why it
+        # should only be here.
+        unless child_of || ignore_active_scope
+          child_of = if scope_manager.active
+                       scope_manager.active.span.context
+                     else
+                       SpanContextFactory.build(datadog_context: datadog_tracer.call_context)
```
Comment (pawelchcki) [inline lib/ddtrace/opentracer/tracer.rb]: 👍

File: spec/ddtrace/opentracer/tracer_integration_spec.rb [hunk 5]
```diff
@@ -204,6 +228,58 @@ def current_trace_for(object)
             end
           end
         end
+
+        context 'preceded by a Datadog span' do
```
Comment (Kyle-Verhoog) [inline spec/ddtrace/opentracer/tracer_integration_spec.rb]: :+1:

File: spec/ddtrace/opentracer/tracer_integration_spec.rb [hunk 6]
```diff
@@ -204,6 +228,58 @@ def current_trace_for(object)
...
+          it { expect(parent_datadog_span.name).to eq(parent_span_name) }
+          it { expect(parent_datadog_span.parent_id).to eq(0) }
+          it { expect(parent_datadog_span.finished?).to be(true) }
+          it { expect(child_datadog_span.name).to eq(span_name) }
+          it { expect(child_datadog_span.parent_id).to eq(parent_datadog_span.span_id) }
+          it { expect(child_datadog_span.finished?).to be(true) }
+          it { expect(child_datadog_span.trace_id).to eq(parent_datadog_span.trace_id) }
+        end
+
+        context 'followed by a Datadog span' do
```
Comment (Kyle-Verhoog) [inline spec/ddtrace/opentracer/tracer_integration_spec.rb]: 👍

File: spec/ddtrace/opentracer/tracer_integration_spec.rb [hunk 7]
```diff
@@ -51,6 +51,30 @@ def current_trace_for(object)
       end
 
       context 'for a nested span' do
+        context 'when there is no active scope' do
```
Comment (Kyle-Verhoog) [inline spec/ddtrace/opentracer/tracer_integration_spec.rb]: 👍

**Pull request merged**